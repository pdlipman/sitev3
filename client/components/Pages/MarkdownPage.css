@import "../../../node_modules/highlight.js/styles/agate.css";
@import "../../../node_modules/github-markdown-css/github-markdown.css";

.markdown-body .highlight pre,
.markdown-body pre {
    padding: 16px;
    overflow: auto;
    font-size: 85%;
    line-height: 1.45;
    background: #333;
    border-radius: 3px;
}

.markdown-body .highlight pre code,
.markdown-body pre code {
    display: block;
    overflow-x: auto;
    padding: 0.5em;
    background: #333;
    color: white;
}