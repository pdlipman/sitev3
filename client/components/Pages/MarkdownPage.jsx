import React from 'react';
import PropTypes from 'prop-types';
import Markdown from 'react-remarkable';
import hljs from 'highlight.js';

import './MarkdownPage.css';

export function highlightCode(codeBlock, language) {
    const renderedCodeBlock = hljs.getLanguage(language)
        ? hljs.highlight(language, codeBlock).value
        : hljs.highlightAuto(codeBlock).value;

    return renderedCodeBlock; // use external default escaping
}

export default class MarkdownPage extends React.Component { // eslint-disable-line react/prefer-stateless-function,max-len
    static propTypes = {
        content: PropTypes.string.isRequired,
    };

    render() {
        const {
            content,
        } = this.props;

        return (
            <div
                className='markdown-body'
            >
                <Markdown
                    options={{
                        typographer: true,
                        quotes: '“”‘’',
                        html: true,
                        xhtmlOut: true, // Use '/' to close single tags (<br />)
                        breaks: true, // Convert '\n' in paragraphs into <br>
                        highlight: highlightCode,
                    }}
                >
                    {content}
                </Markdown>
            </div>
        );
    }
}
