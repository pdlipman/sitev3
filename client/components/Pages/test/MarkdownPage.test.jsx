import React from 'react';
import { shallow } from 'enzyme';

import MarkdownPage, { highlightCode } from '../MarkdownPage.jsx';

describe('Pages/MarkdownPage.jsx', () => {
    const testMarkdown = '### Hello Markdown';
    test('MarkdownPage should render properly', () => {
        const wrapper = shallow(
            <MarkdownPage
                content={testMarkdown}
            />,
        );
        expect(wrapper).toMatchSnapshot();
    });

    test('MarkdownPage highlightCode static function', () => {
        const testCodeBlock = 'console.log("Hello World");';
        const testLanguage = 'javascript';

        expect(highlightCode(testCodeBlock, testLanguage)).toMatchSnapshot();
        expect(highlightCode(testCodeBlock)).toMatchSnapshot();
    });
});
