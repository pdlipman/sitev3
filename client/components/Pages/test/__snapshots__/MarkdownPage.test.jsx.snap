// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pages/MarkdownPage.jsx MarkdownPage highlightCode static function 1`] = `"<span class=\\"hljs-built_in\\">console</span>.log(<span class=\\"hljs-string\\">\\"Hello World\\"</span>);"`;

exports[`Pages/MarkdownPage.jsx MarkdownPage highlightCode static function 2`] = `"console.<span class=\\"hljs-built_in\\">log</span>(<span class=\\"hljs-string\\">\\"Hello World\\"</span>);"`;

exports[`Pages/MarkdownPage.jsx MarkdownPage should render properly 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <MarkdownPage
    content="### Hello Markdown"
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": null,
    "nodeType": "host",
    "props": Object {
      "children": <Remarkable
        container="div"
        options={
          Object {
            "breaks": true,
            "highlight": [Function],
            "html": true,
            "quotes": "“”‘’",
            "typographer": true,
            "xhtmlOut": true,
          }
        }
      >
        ### Hello Markdown
      </Remarkable>,
      "className": "markdown-body",
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": null,
      "nodeType": "class",
      "props": Object {
        "children": "### Hello Markdown",
        "container": "div",
        "options": Object {
          "breaks": true,
          "highlight": [Function],
          "html": true,
          "quotes": "“”‘’",
          "typographer": true,
          "xhtmlOut": true,
        },
      },
      "ref": null,
      "rendered": "### Hello Markdown",
      "type": [Function],
    },
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": null,
      "nodeType": "host",
      "props": Object {
        "children": <Remarkable
          container="div"
          options={
            Object {
              "breaks": true,
              "highlight": [Function],
              "html": true,
              "quotes": "“”‘’",
              "typographer": true,
              "xhtmlOut": true,
            }
          }
        >
          ### Hello Markdown
        </Remarkable>,
        "className": "markdown-body",
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": null,
        "nodeType": "class",
        "props": Object {
          "children": "### Hello Markdown",
          "container": "div",
          "options": Object {
            "breaks": true,
            "highlight": [Function],
            "html": true,
            "quotes": "“”‘’",
            "typographer": true,
            "xhtmlOut": true,
          },
        },
        "ref": null,
        "rendered": "### Hello Markdown",
        "type": [Function],
      },
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
